File: cpp/examples/tic_tac_toe.cpp
#include <boost/functional/hash.hpp>

#include "../gtsa.hpp"

const int SIDE = 3;
const char PLAYER_1 = 'X';
const char PLAYER_2 = 'O';
const char EMPTY = '_';

static char get_opposite_player(char player) {
    return (player == PLAYER_1) ? PLAYER_2 : PLAYER_1;
}

struct TicTacToeMove : public Move<TicTacToeMove> {
    unsigned x;
    unsigned y;

    TicTacToeMove() { }

    TicTacToeMove(unsigned x, unsigned y) : x(x), y(y) { }

    bool operator==(const TicTacToeMove &rhs) const override {
        return x == rhs.x && y == rhs.y;
    }

    void read() override {
        cout << "Enter space separated X and Y of your move: ";
        unsigned x, y;
        cin >> x >> y;
        this->x = x;
        this->y = y;
    }

    ostream &to_stream(ostream &os) const override {
        return os << x << " " << y;
    }
};

const vector<vector<TicTacToeMove>> LINES = [] {
    vector<vector<TicTacToeMove>> lines;
    for (int y = 0; y < SIDE; ++y) {
        lines.emplace_back(); // add a new line to back()
        for (int x = 0; x < SIDE; ++x)
            lines.back().emplace_back(x, y); // add a new coord to that line
    }
    for (int x = 0; x < SIDE; ++x) {
        lines.emplace_back();
        for (int y = 0; y < SIDE; ++y)
            lines.back().emplace_back(x, y);
    }
    lines.emplace_back();
    for (int i = 0; i < SIDE; ++i) {
        lines.back().emplace_back(i, i);
    }
    lines.emplace_back();
    for (int i = 0; i < SIDE; ++i) {
        lines.back().emplace_back(SIDE - i - 1, i);
    }
    return lines;
}();

const unsigned long LINES_SIZE = 2 * SIDE + 2;

struct TicTacToeState : public State<TicTacToeState, TicTacToeMove> {

    vector<char> board;

    TicTacToeState() : State(PLAYER_1) { }

    TicTacToeState(const string &init_string) : State(PLAYER_1) {
        const unsigned long length = init_string.length();
        const unsigned long correct_length = SIDE * SIDE;
        if (length != correct_length) {
            throw invalid_argument("Initialization string length must be " + to_string(correct_length));
        }
        for (int i = 0; i < length; i++) {
            const char c = init_string[i];
            if (c != PLAYER_1 && c != PLAYER_2 && c != EMPTY) {
                throw invalid_argument(string("Undefined symbol used: '") + c + "'");
            }
        }
        board = vector<char>(init_string.begin(), init_string.end());
    }

    TicTacToeState clone() const override {
        TicTacToeState clone = TicTacToeState();
        clone.board = board;
        return clone;
    }

    int get_goodness() const override {
        int goodness = 0;
        const auto &counts = count_players_on_lines(player_to_move);
        for (int i = 0; i < LINES_SIZE; ++i) {
            const int player_places = counts[2 * i];
            const int enemy_places = counts[2 * i + 1];
            if (player_places == SIDE) {
                goodness += SIDE * SIDE;
            }
            else if (enemy_places == SIDE) {
                goodness -= SIDE * SIDE;
            }
            else if (player_places == SIDE - 1 and enemy_places == 0) {
                goodness += SIDE;
            }
            else if (enemy_places == SIDE - 1 and player_places == 0) {
                goodness -= SIDE;
            }
            else if (player_places == SIDE - 2 and enemy_places == 0) {
                ++goodness;
            }
            else if (enemy_places == SIDE - 2 and player_places == 0) {
                --goodness;
            }
        }
        return goodness;
    }

    vector<TicTacToeMove> get_legal_moves() const override {
        vector<TicTacToeMove> result;
        for (unsigned y = 0; y < SIDE; ++y) {
            for (unsigned x = 0; x < SIDE; ++x) {
                if (board[y * SIDE + x] == EMPTY) {
                    result.emplace_back(TicTacToeMove(x, y));
                }
            }
        }
        return result;
    }

    bool is_terminal() const override {
        if (!has_empty_space()) {
            return true;
        }
        const auto &counts = count_players_on_lines(player_to_move);
        for (int i = 0; i < LINES_SIZE; ++i) {
            if (counts[2 * i] == SIDE || counts[2 * i + 1] == SIDE) {
                return true;
            }
        }
        return false;
    }

    bool is_winner(char player) const override {
        const auto &counts = count_players_on_lines(player);
        for (int i = 0; i < LINES_SIZE; ++i) {
            if (counts[2 * i] == SIDE) {
                return true;
            }
        }
        return false;
    }

    void make_move(const TicTacToeMove &move) override {
        board[move.y * SIDE + move.x] = player_to_move;
        player_to_move = get_opposite_player(player_to_move);
    }

    void undo_move(const TicTacToeMove &move) override {
        board[move.y * SIDE + move.x] = EMPTY;
        player_to_move = get_opposite_player(player_to_move);
    }

    bool has_empty_space() const {
        for (unsigned y = 0; y < SIDE; ++y) {
            for (unsigned x = 0; x < SIDE; ++x) {
                if (board[y * SIDE + x] == EMPTY) {
                    return true;
                }
            }
        }
        return false;
    }

    vector<int> count_players_on_lines(char current_player) const {
        vector<int> counts(2 * LINES_SIZE);
        char next_player = get_opposite_player(current_player);
        for (int i = 0; i < LINES_SIZE; ++i) {
            int player_places = 0;
            int enemy_places = 0;
            for (int j = 0; j < SIDE; ++j) {
                const TicTacToeMove &coord = LINES[i][j];
                const int board_index = coord.y * SIDE + coord.x;
                if (board[board_index] == current_player) {
                    ++player_places;
                }
                else if (board[board_index] == next_player) {
                    ++enemy_places;
                }
            }
            counts[2 * i] = player_places;
            counts[2 * i + 1] = enemy_places;
        }
        return counts;
    }

    ostream &to_stream(ostream &os) const override {
        for (int y = 0; y < SIDE; ++y) {
            for (int x = 0; x < SIDE; ++x) {
                os << board[y * SIDE + x];
            }
            os << "\n";
        }
        return os;
    }

    bool operator==(const TicTacToeState &other) const {
        return board == other.board;
    }

    size_t operator()(const TicTacToeState& key) const {
        using boost::hash_value;
        using boost::hash_combine;
        size_t seed = 0;
        hash_combine(seed, hash_value(key.board));
        return seed;
    }
};

File: python/examples/tic_tac_toe.py
from gtsa import State, Human, Tester, Minimax


SIDE = 3
PLAYER_1 = 'X'
PLAYER_2 = 'O'
EMPTY = '_'


def get_opposite_player(player):
    return PLAYER_2 if player == PLAYER_1 else PLAYER_1


LINES = []
for y in range(SIDE):
    row = tuple((x, y) for x in range(SIDE))
    LINES.append(row)
for x in range(SIDE):
    col = tuple((x, y) for y in range(SIDE))
    LINES.append(col)
LINES.append(tuple((x, x) for x in range(SIDE)))
LINES.append(tuple((SIDE - x - 1, x) for x in range(SIDE)))


class TicTacToeState(State):
    def __init__(self, init_string=None):
        super(TicTacToeState, self).__init__(PLAYER_1)

        self.board = [[EMPTY for _ in range(SIDE)] for _ in range(SIDE)]
        if init_string:
            correct_length = SIDE ** 2
            if len(init_string) != correct_length:
                raise ValueError("Initialization string length must be {}".
                                 format(correct_length))
            for i, char in enumerate(init_string):
                if char not in [PLAYER_1, PLAYER_2, EMPTY]:
                    raise ValueError("Undefined symbol used: '{}'".
                                     format(char))
                x = i % SIDE
                y = i // SIDE
                self.board[y][x] = char

    def clone(self):
        clone = TicTacToeState()
        clone.board = [row[:] for row in self.board]
        return clone

    def get_goodness(self):
        goodness = 0
        counts = self.count_players_on_lines(self.player_to_move)
        for count in counts:
            if count[0] == SIDE:
                goodness += SIDE ** 2
            elif count[1] == SIDE:
                goodness -= SIDE ** 2
            elif count[0] == SIDE - 1 and count[1] == 0:
                goodness += SIDE
            elif count[1] == SIDE - 1 and count[0] == 0:
                goodness -= SIDE
            elif count[0] == SIDE - 2 and count[1] == 0:
                goodness += 1
            elif count[1] == SIDE - 2 and count[0] == 0:
                goodness -= 1
        return goodness

    def get_legal_moves(self):
        for y in range(SIDE):
            for x in range(SIDE):
                if self.board[y][x] == EMPTY:
                    yield (x, y)

    def is_terminal(self):
        if not self.has_empty_space():
            return True
        for count in self.count_players_on_lines(self.player_to_move):
            if count[0] == SIDE or count[1] == SIDE:
                return True
        return False

    def is_winner(self, player):
        for count in self.count_players_on_lines(player):
            if count[0] == SIDE:
                return True
        return False

    def make_move(self, move):
        self.board[move[1]][move[0]] = self.player_to_move
        self.player_to_move = get_opposite_player(self.player_to_move)

    def undo_move(self, move):
        self.board[move[1]][move[0]] = EMPTY
        self.player_to_move = get_opposite_player(self.player_to_move)

    def __repr__(self):
        return '\n'.join([''.join(row) for row in self.board]) + '\n'

    def __hash__(self):
        return hash(tuple(tuple(row) for row in self.board))

    def __eq__(self, other):
        return self.board == other.board

    def count_players_on_lines(self, current_player):
        next_player = get_opposite_player(current_player)
        for line in LINES:
            player_places = 0
            enemy_places = 0
            for (x, y) in line:
                if self.board[y][x] == current_player:
                    player_places += 1
                elif self.board[y][x] == next_player:
                    enemy_places += 1
            yield (player_places, enemy_places)

    def has_empty_space(self):
        for y in range(SIDE):
            for x in range(SIDE):
                if self.board[y][x] == EMPTY:
                    return True
        return False


def read_tic_tac_toe_move():
    message = "Enter space separated X and Y of your move: "
    # Both Python 2 and 3 will use raw_input() in place of input().
    try:
        input = raw_input
    except NameError:
        pass
    return tuple(map(int, input(message).split()))


if __name__ == "__main__":
    state = TicTacToeState("___"
                           "___"
                           "___")

    algorithm_1 = Minimax(PLAYER_1, PLAYER_2, verbose=True)
    algorithm_2 = Human(PLAYER_2, PLAYER_1, read_tic_tac_toe_move)

    tester = Tester(state, algorithm_1, algorithm_2)
    tester.start()
